documentation/uniform-versioning.md

```diff
@@ -0,0 +1,95 @@
+# Uniform versioning
+
+> [!NOTE]
+> This article uses terminology defined in the [glossary], like `service` or `service group`.
+
+A `service` **must version uniformly**. 
+
+In brief, it means:
+
+> The service, its set of operation endpoints (typically HTTP REST API endpoints), its documentation, and any SDK referencing it, all must version in lockstep.
+> If a new service API version is released, also a new documentation reference must be released to describe it.
+> Similarly, a new version of given SDK package must be released to refer to the new service version.
+
+Each `service` within a `service group` can version independently of each other.
+
+## Uniform versioning rules
+
+All of the rules listed here apply both for OpenAPI specs emitted from TypeSpec as well as for hand-written OpenAPI specs.
+
+**Uniform versioning** prescribes the following
+
+1. Any deployed service operation endpoint must belong to an API version. An API version once deployed is immutable.
+   Its behavior or constituent operations cannot be changed.
+2. Any given service API version can be composed only of HTTP API operations and ARM resource types (if applicable)
+   that have the same API version. 
+3. Any documentation pertaining to the service and any SDKs generated from the service must pertain to only
+   one service version.
+4. The service version must be always represented in its entirety; any SDK or documentation referring to the service
+   must encompass all of it.
+5. The `common-types` OpenAPI definition shared across multiple services can version independently of the service.
+   However, the rule that API version is immutable still must be observed.
+  As such,  when versioning `common-types`, previous API versions will remain immutable because new API versions must be created.
+
+## Uniform versioning implications
+
+The **uniform versioning** has several implications and implementation decisions supporting it.
+
+### API versions
+
+- The service is effectively defined as a series of consecutive API versions.
+- Each API version is represented by a pair of folders representing its lifecycle stage and a date, 
+  like `stable/2024-03-05` or `preview/2024-05-15-preview`.
+- Each API version date must be later than the previous date.
+- `stable` and `preview` API versions cannot have the same date. This would prevent API users from knowing
```

Comment:
Do we require the newly added version(no matter preview or stable) would be always latter than previous one?

Sometimes service team may have two preview versions e.g `preview/2023-05-15-preview` and `preview/2024-05-15-preview` but they want to stablelize the first one, we may have version `stable/2023-05-16`. Do we allow this?

Reply:
@MaryGao having more than 1 active preview is forbidden. In your case the `preview/2023-05-15-preview` should have been decommissioned and any new `stable` must be a successor to `preview/2024-05-15-preview`. The earliest date it can have is `stable/2024-05-16`.

See also:

[FALSE: My service supports preview features in API versions that came before our current GA](https://eng.ms/docs/products/azure-developer-experience/design/specs-pr-guides/pr-brch-deep#false-my-service-supports-preview-features-in-api-versions-that-came-before-our-current-ga)

@mikekistler and @JeffreyRichter can you double check my guidance here?

Reply:
I concur with this guidance.